### PR TITLE
feat(fwc): add takeoff surveillance 1 warnings and various other improvements

### PR DESCRIPTION
**Pull request by tracernz**
Fixes #[issue_no]

## Summary of Changes

Adds the TOS1 warnings, along with some associated ones whose logic was required, clean up some ADR issues, and makes some improvements to the FWC to allow these warnings to operate correctly.

- **chore(msfssdk): add missing DebounceTimer**
  This was missing from our SDK import
- **fix: clean up adr altitude labels and remove isa deviation**
    Some of the ADR labels are not correct and since I needed to add the altitude label (203) I had to remove the incorrectly labelled baro corrected altitude label... cleaned up the mess a bit. I have not implemented the baro corrected alt labels correctly yet as they still rely on the sim altimeters => they change to pressure alt when STD is selected, but at least the proper labels are there so other subsystems canhook up to the proper labels and end up correct in future.
    There are things that probably ought to be using pressure alt rather than baro corrected alt, but I didn't go into researching all of that. That's a result of having the ADR labels messed up since the beginning.
- **fix(cdu): scratchpad bug**
  Spurious error/race during init that I happened to see during development of this PR
- **fix(fwc): add method to read mtrig**
  It's convenient to be able to read the output like with the other nodes.
- **feat(fwc): implement pulse node**
  Need to be able to have a minimum duration rising or falling edge pulse
- **fix(fwc): split aural and text warnings**
  Not perfect, but at least better than before and enables proper implementation of the warnings in this branch.. Rust FWC will replace all this with proper function.
- **fix(fwc): fix to config/clr/recall**
  Bit closer to proper logic so that the new (and old t.o) warnings can function properly...
- **feat(fwc): takeoff surveillance and associated warnings**
  Implemented the TOS1 and some other warnings that I needed logic from, or are somehow related to the work I was doing here. Also the TOS1 scratchpad messages in the FMGC.

New messages are:
- T.O SPEEDS TOO LOW
  -TOW AND T.O DATA.CHECK
- T.O V1/VR/V2 DISAGREE
- T.O SPEEDS NOT INSERTED
- F/CTL FLAP LVR NOT ZERO
- CONFIG PITCH TRIM
      NOT IN T.O RANGE
- CONFIG
  SPD BRK NOT RETRACTED
- CONFIG RUD TRIM
  NOT IN T.O RANGE
- F/CTL PITCH TRIM/MCDU/CG
       DISAGREE
- F/CTL FLAP/MCDU DISAGREE

Modified messages:
- TO memo
- CONFIG
  SLATS NOT IN T.O CONFIG
- CONFIG
  FLAPS NOT IN T.O CONFIG

## Screenshots (if necessary)

Pressing TO CONF TEST after 1st engine start:
image
Advancing thrust to takeoff power:
image

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Check per the summary above.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Review by RichardPilbery — DISMISSED** [review]

File: fbw-a32nx/src/systems/instruments/src/EWD/PseudoFWC.ts [hunk 1]
```diff
@@ -666,6 +893,57 @@ export class PseudoFWC {
         this.parkBrake.set(SimVar.GetSimVarValue('L:A32NX_PARK_BRAKE_LEVER_POS', 'Bool'));
         this.nwSteeringDisc.set(SimVar.GetSimVarValue('L:A32NX_HYD_NW_STRG_DISC_ECAM_MEMO', 'Bool'));
 
+        /* 22 - AUTOFLIGHT */
+        const fm1DiscreteWord3 = SimVar.GetSimVarValue('L:A32NX_FM1_DISCRETE_WORD_3', 'number');
+        const fm2DiscreteWord3 = SimVar.GetSimVarValue('L:A32NX_FM2_DISCRETE_WORD_3', 'number');
+
+        if (!([2, 3].includes(this.fwcFlightPhase.get()))) {
```
Comment (Eearslya) [inline fbw-a32nx/src/systems/instruments/src/EWD/PseudoFWC.ts:900]: Why not use the convenience variable you just made above?

File: fbw-a32nx/src/systems/instruments/src/EWD/PseudoFWC.ts [hunk 2]
```diff
@@ -666,6 +893,57 @@ export class PseudoFWC {
...
+
+        // TO SPEEDS NOT INSERTED
+        const fmToSpeedsNotInserted = ((fm1DiscreteWord3 & fm2DiscreteWord3 & 1 << 18) > 0);
+
+        this.toConfigAndNoToSpeedsPulseNode.write(fmToSpeedsNotInserted && this.toConfigTestRaw, deltaTime);
+
+        if (fmToSpeedsNotInserted && (this.toConfigTestRaw || this.fwcFlightPhase.get() === 3) && !this.toSpeedsNotInserted) {
+            this.toSpeedsNotInserted = true;
+        }
+        if (!([2, 3].includes(this.fwcFlightPhase.get()) && fmToSpeedsNotInserted) && this.toSpeedsNotInserted) {
```
Comment (Eearslya) [inline fbw-a32nx/src/systems/instruments/src/EWD/PseudoFWC.ts:914]: Ditto above.

**Review by Eearslya — APPROVED** [review]
Minor nitpick, otherwise code looks good to me.

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Utils/NXSpeeds.js [hunk 3]
```diff
@@ -560,4 +586,49 @@ class NXSpeedsUtils {
...
+    static interpolateTable(table, alt) {
+        if (alt <= table[0][0]) {
+            return vmca[0][1];
+        }
+        if (alt >= table[table.length - 1][0]) {
+            table[table.length - 1][1];
+        }
+        for (let i = 0; i < table.length - 1; i++) {
+            if (alt >= table[i][0] && alt <= table[i + 1][0]) {
+                const d = (alt - table[i + 1][0]) / (table[i + 1][0] - table[i][0]);
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Utils/NXSpeeds.js:600]: Is it intentional that `d` will be negative here?

**Review by BlueberryKing — CHANGES_REQUESTED** [review]

File: fbw-a32nx/src/systems/instruments/src/EWD/PseudoFWC.ts (hunk 1 above)
Comment (tracernz) [reply fbw-a32nx/src/systems/instruments/src/EWD/PseudoFWC.ts:900]: Good catch. I did do some search/replace of each incantation, but probably got distracted halfway through by something else I saw 🤣 .

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 4]
```diff
@@ -852,6 +862,26 @@ class FMCMainDisplay extends BaseAirliners {
         return ((288.15 - 0.0019812 * alt + isaDev) / 288.15) ** 5.25588 * 1013.2;
     }
 
+    getPressureAltAtElevation(elev, qnh = 1013.2) {
+        const p0 = qnh < 500 ? 29.92 : 1013.2;
+        return elev + 145442.15 * Math.pow((1 - (qnh / p0)), 0.190263);
```
[the remark this answers is not in the document]
Comment (tracernz) [reply fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js:867]: Oof. Great catch!

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Utils/NXSpeeds.js (hunk 3 above)
Comment (tracernz) [reply fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Utils/NXSpeeds.js:600]: 😨

**Review by BlueberryKing — APPROVED** [review]
Nice work! Heads-up, I didn't check the entire FWC logic for correctness. I think we can do that either through QA or through a future review.

**Comment by 2hwk** [comment]
Fix conflicts ^

**Comment by mico975** [comment]
Quality Assurance Tester Report

Discord          : mico#3145
Object of testing: #7652
Tier of Testing  : 1
Date             : 28/02/2023

Testing Process: 
Went through the summary and tested accordingly, all good.

Negatives:       
N/A

Testing Results: 
Passed

Conclusions: 
LGTM

**Pull request merged**